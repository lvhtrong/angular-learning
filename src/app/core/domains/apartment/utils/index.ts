import { Apartment as StoreApartment } from '@store/state/domains/apartment/apartment.model';
import { ApartmentDTO } from '../models/apartment.dto';

export function mapToStoreApartment(apartment: ApartmentDTO): StoreApartment {
  return {
    id: apartment.id,
    updatedAt: apartment.updatedAt,
    description: apartment.description,
    price: apartment.price,
    address: apartment.address,
    location: apartment.location,
    specs: apartment.specs,
    images: apartment.images,
    utilities: apartment.utilities,
    banner: apartment.banner,
    title: apartment.title,
    type: apartment.type,
    createdBy: apartment.createdBy,
    createdAt: apartment.createdAt,
    roomsCount: apartment.roomsCount,
    rooms: apartment.rooms,
  };
}

export function mapFromStoreApartment(apartment: StoreApartment): ApartmentDTO {
  return {
    id: apartment.id,
    updatedAt: apartment.updatedAt,
    description: apartment.description,
    price: apartment.price,
    address: apartment.address,
    location: apartment.location,
    specs: apartment.specs,
    images: apartment.images,
    utilities: apartment.utilities,
    banner: apartment.banner,
    title: apartment.title,
    type: apartment.type,
    createdBy: apartment.createdBy,
    createdAt: apartment.createdAt,
    roomsCount: apartment.roomsCount,
    rooms: apartment.rooms,
  };
}
