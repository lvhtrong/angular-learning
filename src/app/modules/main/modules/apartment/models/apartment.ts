export interface Apartment {
  id: string;
  title?: string;
  type?: string;
  shortDescription?: string;
  address?: string;
}
