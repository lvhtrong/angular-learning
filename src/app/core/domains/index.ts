import { AuthService } from './auth/services/auth.service';
import { ApartmentService } from './apartment/services/apartment.service';

export const services = [AuthService, ApartmentService];
