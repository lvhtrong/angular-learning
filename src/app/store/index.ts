import {
  ActionReducer,
  ActionReducerMap,
  createFeatureSelector,
  createSelector,
  MetaReducer,
} from '@ngrx/store';
import { environment } from '@env';
import * as fromDomains from './state/domains/domains.reducer';

export interface State {
  [fromDomains.domainsFeatureKey]: fromDomains.State;
}

export const reducers: ActionReducerMap<State> = {
  [fromDomains.domainsFeatureKey]: fromDomains.reducer,
};

export const metaReducers: MetaReducer<State>[] = !environment.production
  ? []
  : [];
