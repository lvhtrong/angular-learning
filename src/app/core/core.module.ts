import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';

import * as fromServices from './services';
import { SharedModule } from '@shared/shared.module';

@NgModule({
  declarations: [],
  imports: [CommonModule, HttpClientModule, SharedModule],
  exports: [],
  providers: [...fromServices.services],
})
export class CoreModule {}
