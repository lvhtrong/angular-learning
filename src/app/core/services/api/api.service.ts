import { Injectable } from '@angular/core';
import { HttpClient, HttpRequest } from '@angular/common/http';
import { RequestOptions } from './models/request-options';

@Injectable({
  providedIn: 'root',
})
export class ApiService {
  constructor(private http: HttpClient) {}

  /**
   * get
   */
  public get<T>(url: string, options?: RequestOptions) {
    return this.http.get<T>(url, options);
  }

  /**
   * post
   */
  public post<T>(url: string, body: any, options?: RequestOptions) {
    return this.http.post<T>(url, body, options);
  }
}
