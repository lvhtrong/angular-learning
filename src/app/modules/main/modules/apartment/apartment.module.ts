import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { ApartmentRoutingModule } from './apartment-routing.module';
import { ListComponent } from './pages/list/list.component';

@NgModule({
  declarations: [ListComponent],
  imports: [CommonModule, ApartmentRoutingModule],
})
export class ApartmentModule {}
