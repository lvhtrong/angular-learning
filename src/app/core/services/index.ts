import { AuthService } from './auth/auth.service';
import { ApiService } from './api/api.service';

export const services = [AuthService, ApiService];
