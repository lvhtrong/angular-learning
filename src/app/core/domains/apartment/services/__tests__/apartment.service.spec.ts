import { TestBed } from '@angular/core/testing';
import { cold } from 'jest-marbles';
import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { ApiService } from '@core/services/api/api.service';
import { ApartmentService } from '../apartment.service';
import { ApartmentDTO } from '../../models/apartment.dto';
import { Observable } from 'rxjs';
import { Apartment } from '@store/state/domains/apartment/apartment.model';

jest.mock('@store/state/domains/apartment/apartment.selectors', () => ({
  selectApartmentEntities: jest.fn((a) => a),
}));
jest.mock('@core/services/api/api.service');

const getApartmentDTO = (id: string): ApartmentDTO => {
  return {
    description: {
      long: 'long-desc',
      short: 'short-desc',
    },
    price: {
      min: 1,
      max: 2,
    },
    address: {
      city: 'city-1',
      district: 'district-1',
      street: 'street-1',
    },
    location: {
      lat: 'lat-1',
      long: 'long-1',
    },
    specs: ['specs-1'],
    images: ['utilities-1'],
    utilities: ['utilities-1'],
    banner: 'banner',
    title: 'title',
    type: 'type',
    createdBy: 'createdBy',
    createdAt: 'createdAt',
    updatedAt: 'updatedAt',
    roomsCount: 2,
    id,
    rooms: ['room1', 'room2'],
  };
};

const getApartmentStore = (id: string): Apartment => {
  return {
    description: {
      long: 'long-desc',
      short: 'short-desc',
    },
    price: {
      min: 1,
      max: 2,
    },
    address: {
      city: 'city-1',
      district: 'district-1',
      street: 'street-1',
    },
    location: {
      lat: 'lat-1',
      long: 'long-1',
    },
    specs: ['specs-1'],
    images: ['utilities-1'],
    utilities: ['utilities-1'],
    banner: 'banner',
    title: 'title',
    type: 'type',
    createdBy: 'createdBy',
    createdAt: 'createdAt',
    updatedAt: 'updatedAt',
    roomsCount: 2,
    id,
    rooms: ['room1', 'room2'],
  };
};

const configureModules = (
  apiServiceValue = {
    get: jest.fn().mockReturnValue(
      new Observable<ApartmentDTO[]>((observer) => {
        observer.next([]);
      })
    ),
  }
) => {
  TestBed.configureTestingModule({
    providers: [
      ApartmentService,
      {
        provide: ApiService,
        useValue: apiServiceValue,
      },
      provideMockStore({
        initialState: [],
      }),
    ],
  });
};

describe('getApartments', () => {
  it('should upsert returned apartments to store', (done) => {
    const apiApartments = [getApartmentDTO('id-1')];
    const apiServiceMockValue = {
      get: jest.fn().mockReturnValue(
        new Observable<ApartmentDTO[]>((observer) => {
          observer.next(apiApartments);
        })
      ),
    };
    configureModules(apiServiceMockValue);

    const store = TestBed.inject(MockStore);
    store.dispatch = jest.fn();

    const apartmentService = TestBed.inject(ApartmentService);
    apartmentService.getApartments().subscribe(() => {
      expect(store.dispatch as jest.Mock).toHaveBeenCalledTimes(1);
      expect((store.dispatch as jest.Mock).mock.calls[0]).toMatchSnapshot();

      done();
    });

    expect(apiServiceMockValue.get).toHaveBeenCalledTimes(1);
    expect(apiServiceMockValue.get).toHaveBeenCalledWith('/apartments');
  });

  it('should map the stored apartment to apartment$ variable', () => {
    configureModules();

    const store = TestBed.inject(MockStore);
    const apartmentService = TestBed.inject(ApartmentService);

    const apartment = getApartmentStore('id-1');
    store.setState([apartment]);

    const expected = cold('a', {
      a: [apartment],
    });

    expect(apartmentService.apartments$).toBeObservable(expected);
  });
});
