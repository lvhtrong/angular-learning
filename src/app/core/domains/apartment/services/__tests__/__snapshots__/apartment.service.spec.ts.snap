// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getApartments should upsert returned apartments to store 1`] = `
Array [
  Object {
    "apartments": Array [
      Object {
        "address": Object {
          "city": "city-1",
          "district": "district-1",
          "street": "street-1",
        },
        "banner": "banner",
        "createdAt": "createdAt",
        "createdBy": "createdBy",
        "description": Object {
          "long": "long-desc",
          "short": "short-desc",
        },
        "id": "id-1",
        "images": Array [
          "utilities-1",
        ],
        "location": Object {
          "lat": "lat-1",
          "long": "long-1",
        },
        "price": Object {
          "max": 2,
          "min": 1,
        },
        "rooms": Array [
          "room1",
          "room2",
        ],
        "roomsCount": 2,
        "specs": Array [
          "specs-1",
        ],
        "title": "title",
        "type": "type",
        "updatedAt": "updatedAt",
        "utilities": Array [
          "utilities-1",
        ],
      },
    ],
    "type": "[Apartment/API] Upsert Apartments",
  },
]
`;
